clients/gtest/rocblas_test.cpp

```diff
@@ -0,0 +1,166 @@
+
+/*********************************************
+ * Signal-handling for detecting test faults *
+ *********************************************/
+
+// Type of sigsetjmp return value
+enum
+{
+    fallthrough = 0,
+    signal_received,
+    terminated,
+};
+
+// State of the signal handler
+static struct
+{
+    // Whether sigjmp_buf is set and catching signals is enabled
+    volatile sig_atomic_t enabled;
+
+    // Id of the thread which is catching signals
+    volatile pthread_t tid;
+
+    // sigjmp_buf describing stack frame to go back to
+    sigjmp_buf sigjmp_buf;
+
+    // The signal which was received
+    volatile int signal;
+} handler;
+
+// Signal handler (must have external "C" linkage)
+extern "C" void rocblas_test_signal_handler(int sig)
+{
+    int saved_errno = errno; // Save errno
+
+    // If this is an alarm timeout, we abort
+    if(sig == SIGALRM)
+    {
+        static constexpr char msg[]
+            = "\nAborting tests due to an alarm timeout.\n\n"
+              "This could be due to a deadlock caused by mutexes being left locked\n"
+              "after a previous test's signal was caught and partially recovered from.\n";
+        write(STDERR_FILENO, msg, sizeof(msg) - 1);
+        rocblas_abort();
+    }
+
+    // If the signal handler is disabled, because we're not in the middle of
+    // running a rocBLAS test, restore this signal's disposition to default,
+    // and reraise the signal
+    if(!handler.enabled)
+    {
+        signal(sig, SIG_DFL);
+        errno = saved_errno;
+        raise(sig);
+        return;
+    }
+
+    // If the thread receiving the signal is different from the thread
+    // catching signals, send the signal to the thread catching signals.
+    if(!pthread_equal(pthread_self(), handler.tid))
+    {
+        pthread_kill(handler.tid, sig);
+        errno = saved_errno;
+        return;
+    }
+
+    // Jump back to the handler code after setting handler.signal
+    // Note: This bypasses stack unwinding, and may lead to memory leaks, but
+    // it is better than crashing.
+    handler.signal = sig;
+    errno          = saved_errno;
+    siglongjmp(handler.sigjmp_buf, signal_received);
+}
+
+// Catch std::terminate
+static void terminate_handler()
+{
+    if(handler.enabled)
+        siglongjmp(handler.sigjmp_buf, terminated);
+    std::set_terminate(nullptr);
+    std::terminate();
+}
+
+// Set up signal handlers
+void rocblas_test_sigaction()
+{
+    struct sigaction act;
+    sigfillset(&act.sa_mask);
+    act.sa_flags   = 0;
+    act.sa_handler = rocblas_test_signal_handler;
+
+    for(int sig : {SIGABRT, SIGBUS, SIGFPE, SIGILL, SIGALRM, SIGSEGV, SIGSYS, SIGUSR1, SIGUSR2})
+        sigaction(sig, &act, nullptr);
+}
+
+// Number of seconds each test is allowed to take before all testing is killed.
+// If this time limit is exceeded for any particular test, it is assumed that
+// there is a permanent deadlock, and testing is terminated.
+constexpr int TEST_TIMEOUT = 600;
+
```

Comment:
Often CI sets testing timeouts so not sure if we want to expose to them as environment variable.  E.g. simulator would require huge timeout.

Reply:
The `ROCBLAS_TEST_TIMEOUT` environment variable has been added to allow you to override the default of `600` seconds.

Reply:
If you set `ROCBLAS_TEST_TIMEOUT=0`, you will disable the alarm entirely.